src/ivc/incrementally_verifiable_computation.rs

```diff
@@ -228,17 +231,89 @@ where
 
     pub fn verify<const T: usize, RP1, RP2>(
         &mut self,
-        _pp: &PublicParams<'_, A1, A2, T, C1, C2, SC1, SC2, RP1, RP2>,
-        _steps_count: usize,
-        _z0_primary: [C1::Scalar; A1],
-        _z0_secondary: [C2::Scalar; A2],
+        pp: &PublicParams<'_, A1, A2, T, C1, C2, SC1, SC2, RP1, RP2>,
+        num_steps: usize,
+        z0_primary: [C1::Scalar; A1],
+        z0_secondary: [C2::Scalar; A2],
     ) -> Result<(), Error>
     where
-        RP1: ROPair<C1::Scalar>,
-        RP2: ROPair<C2::Scalar>,
+        RP1: ROPair<C1::Scalar, Config = MainGateConfig<T>>,
+        RP2: ROPair<C2::Scalar, Config = MainGateConfig<T>>,
     {
-        // TODO #31
-        todo!("Logic at `RecursiveSNARK::verify`")
+        let is_step_zero = num_steps == 0;
+        let is_num_step_not_match = num_steps != self.step;
+        let is_inputs_match = self.primary.z_0 != z0_primary || self.secondary.z_0 != z0_secondary;
+        if is_step_zero || is_num_step_not_match || is_inputs_match {
+            return Err(Error::IVCParamNotMatch);
+        }
+
+        // verify X0
+        let ro1 = &mut RP1::OffCircuit::new(pp.primary.params.ro_constant.clone());
+        ro1.absorb_point(&pp.digest::<C2>().unwrap());
+        ro1.absorb_field(C1::Scalar::from_u128(num_steps as u128));
+        z0_primary.iter().for_each(|val| {
+            ro1.absorb_field(*val);
+        });
+        self.primary
+            .z_i
+            .iter()
+            .for_each(|val| ro1.absorb_field(*val));
+        self.secondary.relaxed_trace.U.absorb_into(ro1);
+        let _hash_primary = ro1.squeeze::<C2>(NUM_CHALLENGE_BITS);
+        // TODO: uncomment after adding secondary_trace field in IVC
+        // assert_eq!(hash_primary, self.secondary_trace.U.X0);
+
+        // verify X1
+        let ro2 = &mut RP2::OffCircuit::new(pp.secondary.params.ro_constant.clone());
+        ro2.absorb_point(&pp.digest::<C1>().unwrap());
+        ro2.absorb_field(C2::Scalar::from_u128(num_steps as u128));
+        z0_secondary.iter().for_each(|val| {
+            ro2.absorb_field(*val);
+        });
+        self.secondary
+            .z_i
+            .iter()
+            .for_each(|val| ro2.absorb_field(*val));
+        self.primary.relaxed_trace.U.absorb_into(ro2);
+        let _hash_secondary = ro2.squeeze::<C1>(NUM_CHALLENGE_BITS);
+        // TODO: uncomment after adding secondary_trace field in IVC
+        // assert_eq!(hash_secondary, self.secondary_trace.U.X1);
+
+        // check satisfiability
+        // TODO: avoid create PlonkStructure again
+        let (primary_td, _) = Self::prepare_primary_td::<T, RP1>(
+            pp.primary.k_table_size,
+            [C1::Scalar::ZERO, C1::Scalar::ZERO],
```

Comment:
Add `TODO` here like above

Reply:
you mean the X0 and X1? Since we only need PlonkStructure, we can use ZERO here.

Reply:
So once again, back to the context of #154. We can't do inputize as easily as Nova, we have to honestly count these values and pass them here, because otherwise it comes out that they are not instance columns. 

Please let's discuss in issue and come to a consensus, there seems to be a misunderstanding here

Reply:
yeah, there is some misunderstanding. What I mean is:
In order to call `is_sat` and `is_sat_relaxed`, we need have get PlonkStructure first. However, the checking of `is_sat` and `is_sat_relaxed` do not involve `X0` and X1`, hence we can just use any number here without affect the actually checking